wgsl/index.bs

```diff
@@ -807,193 +975,83 @@ then:
-type |S| in storage class |C|, where |C| is either [=storage classes/storage=] or [=storage classes/storage=].
-It is defined recursively in the following table:
+In this section we define <dfn noexport>RequiredAlignOf</dfn>(|S|, |C|) as the
+required alignment of host-shareable type |S| when used by storage class |C|.
 
 <table class='data'>
-  <caption>Alignment of a host-shareable type</caption>
+  <caption>
+    Alignment requirements of a host-shareable type for
+    [=storage classes/storage=] and [=storage classes/uniform=] storage classes
+  </caption>
   <thead>
     <tr><th>Host-shareable type |S|
-        <th>|Align|(|S|,`storage`)
-        <th>|Align|(|S|,`uniform`)
+        <th>[=RequiredAlignOf=](|S|, [=storage classes/storage=])
+        <th>[=RequiredAlignOf=](|S|, [=storage classes/uniform=])
   </thead>
   <tr><td>[=i32=], [=u32=], or [=f32=]
-      <td>4
-      <td>4
-  <tr><td>vec2&lt;|T|&gt;,
-      where |T| is one of [=i32=], [=u32=], or [=f32=]
-      <td>8
-      <td>8
-  <tr><td>vec3&lt;|T|&gt;,
-      where |T| is one of [=i32=], [=u32=], or [=f32=]
-      <td>16
-      <td>16
-  <tr><td>vec4&lt;|T|&gt;,
-      where |T| is one of [=i32=], [=u32=], or [=f32=]
-      <td>16
-      <td>16
+      <td>[=AlignOf=](|S|)
+      <td>[=AlignOf=](|S|)
+  <tr><td>vec|N|&lt;`T`&gt;
+      <td>[=AlignOf=](|S|)
+      <td>[=AlignOf=](|S|)
   <tr algorithm="alignment of a matrix with 2 rows">
-      <td>mat|N|x2&lt;f32&gt;
-      <td>8
-      <td>8
-          <!-- Vulkan base alignment: 2 times scalar aligment -->
-  <tr algorithm="alignment of a matrix with 3 rows">
-      <td>mat|N|x3&lt;f32&gt;
-      <td>16
-      <td>16
-          <!-- Vulkan base alignment: 4 times scalar aligment -->
-  <tr algorithm="alignment of a matrix with 4 rows">
-      <td>mat|N|x4&lt;f32&gt;
-      <td>16
-      <td>16
-          <!-- Vulkan base alignment: 4 times scalar aligment -->
+      <td>mat|N|x|M|&lt;f32&gt;
+      <td>[=AlignOf=](|S|)
+      <td>[=AlignOf=](|S|)
   <tr algorithm="alignment of an array">
       <td>array<|T|,|N|>
-      <td>|Align|(|T|,`storage`)
-      <td>[=roundUp=](16, |Align|(|T|,`uniform`))
+      <td>[=AlignOf=](|T|)
+      <td>[=roundUp=](16, [=AlignOf=](|T|))
   <tr algorithm="alignment of an runtime-sized array">
       <td>array<|T|>
-      <td>|Align|(|T|,`storage`)
-      <td>[=roundUp=](16, |Align|(|T|,`uniform`))
+      <td>[=AlignOf=](|T|)
+      <td>[=roundUp=](16, [=AlignOf=](|T|))
   <tr algorithm="alignment of a structure">
-      <td>struct&lt;|T1|,...,|Tn|&gt;
-      <td>max(|Align|(|T1|,`storage`),..., |Align|(|Tn|,`storage`))
-      <td>[=roundUp=](16, |A|),<br> where |A| = max(|Align|(|T1|,`uniform`),..., |Align|(|Tn|,`uniform`)))
+      <td>struct&lt;T<sub>0</sub>, ..., T<sub>N</sub>&gt;
+      <td>max([=AlignOf=](T<sub>0</sub>), ..., [=AlignOf=](T<sub>N</sub>))
+      <td>[=roundUp=](16, max([=AlignOf=](T<sub>0</sub>), ..., [=AlignOf=](T<sub>N</sub>)))<br>
```

Comment:
I know I suggested this way, but looking at it again I wonder if defining it in terms of storage is better (for uniform)
```
roundUp(16, RequiredAlignOf(S, storage))
```

I'm not sure it is, but I leave it to your discretion.

Reply:
I'm not sure any more.

This had me scratching my head a bit to make sure the values were actually identical to what was there before, but as `16` bytes is the largest required alignment of any type, I think this is the same.

In terms of reading the spec, I feel what we currently have is easier to understand, but I don't know if we're folding away too much that'll bite us if / when rules change.